src/main/java/com/arpnetworking/tsdcore/sinks/WavefrontSink.java

```diff
@@ -0,0 +1,357 @@
+            _purgeBuffer = value;
+            return self();
+        }
+
+        /**
+         * Milliseconds between flushes.  Defaults to 1000ms.
+         *
+         * @param value Milliseconds between flushes.
+         * @return This instance of <code>Builder</code>
+         */
+        public Builder setPushFlushInterval(final Long value) {
+            _pushFlushInterval = value;
+            return self();
+        }
+
+        /**
+         * Maximum allowed points in a single push flush. Defaults to 50,000.
+         *
+         * @param value Maximum points allowed per flush.
+         * @return This instance of <code>Builder</code>
+         */
+        public Builder setPushFlushMaxPoints(final Integer value) {
+            _pushFlushMaxPoints = value;
+            return self();
+        }
+
+        /**
+         * File to read agent id from. Cannot be null or empty.
+         *
+         * @param value Id file name.
+         * @return This instance of <code>Builder</code>
+         */
+        public Builder setIdFile(final String value) {
+            _idFile = value;
+            return self();
+        }
+
+        /**
+         * Set of dimension names to be passed through as a Wavefront annotation/tag.
+         *
+         * @param value Set of white listed dimension names
+         * @return This instance of <code>Builder</code>
+         */
+        public Builder setWhiteListedDimensions(final Set<String> value) {
+            _whiteListedDimensions = new ImmutableSet.Builder<String>().addAll(value).build();
+            return self();
+        }
+
+        /**
+         * Base name for Wavefront agent buffer files. Defaults to 'buffer' if not specified.
+         *
+         * @param value Base name of buffer files.
+         * @return This instance of <code>Builder</code>
+         */
+        public Builder setBufferBasename(final String value) {
+            _bufferBasename = value;
+            return self();
+        }
+
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected Builder self() {
+            return this;
+        }
+
+        @NotNull
+        private URI _uri;
+        @NotNull
+        @NotEmpty
+        private String _token;
+        private Integer _retryThreads;
+        private Integer _flushThreads;
+        @NotNull
+        private Boolean _purgeBuffer = false;
+        private Long _pushFlushInterval;
+        private Integer _pushFlushMaxPoints;
+        @NotNull
+        @NotEmpty
+        private String _idFile;
+        @NotNull
+        private ImmutableSet<String> _whiteListedDimensions = new ImmutableSet.Builder<String>().build();
+        private String _bufferBasename;
+    }
+
+    private static final class AgentFactory {
+        private static final Map<String, WavefrontAgent> AGENT_MAP = Maps.newHashMap();
+
+        public static synchronized WavefrontAgent getInstance(final String token, final List<String> arguments) {
```

Comment:
Would this be more efficient as a concurrent hash map with computeIfAbsent?

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html